transformer_engine/jax/sharding.py

```diff
@@ -1015,6 +1033,98 @@ def get_fused_attn_sharding_meta(stype: ShardingType,
                                                               tp_axis_name)
 
 
+def extend_fsdp_sharding_meta(sharding_meta: ShardingMeta,
+                              input_dp_dim: int = 0,
+                              weight_fsdp_dim_map: Dict[int, int] = None) -> ShardingMeta:
+    """
+    extend_fsdp_sharding_meta
+    """
+    weight_fsdp_dim_map = {} if weight_fsdp_dim_map is None else weight_fsdp_dim_map
+
+    mst = infer_major_sharding_type()
+    if mst is MajorShardingType.SINGLE:
+        return sharding_meta, ""
+
+    gsr = global_shard_resource()
+    dp_mesh_axis = gsr.dp_resource
+    fsdp_mesh_axis = gsr.fsdp_resource
+
+    if fsdp_mesh_axis == dp_mesh_axis:
+        return sharding_meta, ""
+    if fsdp_mesh_axis is None:
+        return sharding_meta, ""
+
+    fsdp_dim_size, _ = _get_mesh_info(fsdp_mesh_axis)
+    fsdp_axis_name = "fsdp"
+
+    def get_idx_to_extend(sharded_indices, target_idx):
+        idx_to_extend = target_idx
+        for i in sharded_indices:
+            if i <= target_idx:
+                idx_to_extend += 1
+        return idx_to_extend
+
+    def extend_exist_sharding(idx, shape):
+        remain_size = shape[idx]
+        assert remain_size == -1 or remain_size % fsdp_dim_size == 0
+        remain_size = remain_size // fsdp_dim_size
+        new_shape = tuple([*shape[:idx], fsdp_dim_size, remain_size, *shape[idx + 1:]])
+        return new_shape
+
+    new_input_shapes = []
+    new_in_axes = []
+    for i, shape in enumerate(sharding_meta.input_shapes):
+        idx_to_extend = -1
+        if i == 0:    # Assume 0-idx in the tensor with inputs
```

Comment:
Done:
1. Changed the comment to `first shape corresponds to input`.
2. Added a documentation.

About mergeing `input_dp_dim` into `weight_fsdp_dim_map`:
We prefer to keep `input_dp_dim`, because it kind of force to set the dimension of input to extend FSDP. If we moving that into `weight_fsdp_dim_map`, users might forget to set `{0:xxx}` then led to errors. Instead of removing `input_dp_dim`, we added a assertion to remind users that `0` should not be in `weight_fsdp_dim_map`.
Does this make sense?

Reply:
The documentation is much better. I'm still wary of treating `input_dp_dim` and `weight_fsdp_dim_map` separately though, since the added convenience comes at the cost of making the code harder to understand and generalize. If we just had `weight_fsdp_dim_map`, it seems we could remove the branching behavior at: https://github.com/NVIDIA/TransformerEngine/blob/6f071004edeeb62e43eb77b432fe56fffc179a45/transformer_engine/jax/sharding.py#L1101

Two possible ideas:
1. Remove `weight_fsdp_dim_map` and change the assert to something like:
```python
assert 0 in weight_fsdp_dim_map, "Batch dimension for input tensor was not provided."
```
2. If you insist on the current API, you can simplify the implementation by incorporating `input_dp_dim` into `weight_fsdp_dim_map`:
```python
assert 0 not in weight_fsdp_dim_map, "Index 0 is reserved for the input."
weight_fsdp_dim_map[0] = input_dp_dim
```

Reply:
I see. Merged `input_dp_dim` into `weight_fsdp_dim_map` and added a proper assertion.
Regarding the branch of `if i == 0:`, currently there is a bit differenence of extending FSDP to input tensors and weights, then it could not be directly removed, but agree that they are similar. We have an internal plan to migrate current sharding approaches to new JAX feature, called `custom_partitioning`. Then we will refactor at that migration to make code being more clear.